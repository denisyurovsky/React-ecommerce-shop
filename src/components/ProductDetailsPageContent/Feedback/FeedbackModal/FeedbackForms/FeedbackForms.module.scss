@import "../../../../../assets/styles/colors.scss";

.checkbox {
  margin-bottom: 20px;

  & span {
    font-size: 1.2rem;
    color: $greyDark;
  }
}
