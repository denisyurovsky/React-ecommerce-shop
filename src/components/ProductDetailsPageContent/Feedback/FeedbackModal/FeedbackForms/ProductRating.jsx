import StarRateIcon from '@mui/icons-material/StarRate';
import { Box, Typography, FormLabel, Rating } from '@mui/material';
import PropTypes from 'prop-types';
import React, { useState } from 'react';

import styles from './productRating.module.scss';

const labels = {
  1: 'Awful',
  2: 'Bad',
  3: 'Ok',
  4: 'Good',
  5: 'Excellent',
};

const ProductRating = ({ value, onChange }) => {
  const [hover, setHover] = useState(-1);

  const handleChange = (_, newValue) => onChange(newValue);
  const handleHover = (_, newValue) => setHover(newValue);

  const getLabelText = (value) => `${value} Star${value !== 1 ? 's' : ''}`;

  return (
    <>
      <FormLabel component="legend">Rating</FormLabel>
      <Box className={styles.container}>
        <Rating
          name="hover-feedback"
          value={value}
          onChange={handleChange}
          onChangeActive={handleHover}
          emptyIcon={
            <StarRateIcon style={{ opacity: 0.55 }} fontSize="inherit" />
          }
          getLabelText={getLabelText}
          data-cy="starRating"
        />
        <Typography component="legend" sx={{ ml: 2 }}>
          {value !== null && labels[hover !== -1 ? hover : value]}
        </Typography>
      </Box>
    </>
  );
};

ProductRating.propTypes = {
  onChange: PropTypes.func.isRequired,
  value: PropTypes.number.isRequired,
};

export default ProductRating;
