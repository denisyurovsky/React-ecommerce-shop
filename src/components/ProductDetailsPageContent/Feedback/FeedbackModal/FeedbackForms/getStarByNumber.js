import { screen } from '@testing-library/react';

const getStarByNumber = (number) => {
  const name = `${number} Star${number > 1 ? 's' : ''}`;

  return screen.getByRole('radio', { name });
};

export default getStarByNumber;
