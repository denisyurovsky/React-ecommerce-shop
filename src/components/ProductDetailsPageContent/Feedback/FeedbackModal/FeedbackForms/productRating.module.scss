.container {
  display: flex;
  margin-bottom: 25px;
}
