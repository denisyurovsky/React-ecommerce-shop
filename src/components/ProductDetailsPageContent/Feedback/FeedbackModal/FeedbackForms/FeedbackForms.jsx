import { TextField, Box, Button, Checkbox } from '@mui/material';
import FormControlLabel from '@mui/material/FormControlLabel';
import PropTypes from 'prop-types';
import React, { useState } from 'react';
import { useSelector } from 'react-redux';

import {
  MIN_COMMENT_LENGTH,
  DEFAULT_NAME,
} from '../../../../../constants/feedbackConstants.js';
import { getUser } from '../../../../../store/user/userSlice';
import LoadingButton from '../../../../ui-kit/buttons/LoadingButton.jsx';

import ProductRating from './ProductRating';

import styles from './FeedbackForms.module.scss';

const FeedbackForms = ({ sendForm, isLoading }) => {
  const [userName, setUserName] = useState('');
  const [comment, setComment] = useState('');
  const [rating, setRating] = useState(-1);
  const [isFinished, setIsFinished] = useState(false);
  const [isRealUserName, setIsRealUserName] = useState(false);
  const user = useSelector(getUser);

  const checkForm = (rating, comment) =>
    checkRating(rating) && checkComment(comment);
  const checkRating = (value) => value > 0;
  const checkComment = (text) => text.length >= MIN_COMMENT_LENGTH;

  const handleNameChange = (evt) => setUserName(evt.target.value);
  const handleCommentChange = (evt) => {
    const text = evt.target.value;

    setIsFinished(checkForm(rating, text));
    setComment(text);
  };
  const handleChange = (value) => {
    setIsFinished(checkForm(value, comment));
    setRating(value);
  };

  const prepareToSend = () => {
    const name = userName === '' ? DEFAULT_NAME : userName;

    sendForm({ name, rating, comment });
  };

  const checkboxHandler = (e) => {
    e.target.checked
      ? setUserName(`${user.firstName} ${user.lastName}`)
      : setUserName(DEFAULT_NAME);

    setIsRealUserName(!isRealUserName);
  };

  return (
    <Box>
      <TextField
        value={userName}
        onChange={handleNameChange}
        autoComplete="off"
        label="Enter your name"
        placeholder={DEFAULT_NAME}
        variant="standard"
        fullWidth
        disabled={isRealUserName}
      />
      <FormControlLabel
        control={
          <Checkbox
            onChange={checkboxHandler}
            checked={isRealUserName}
            inputProps={{ 'aria-label': 'pasteName' }}
          />
        }
        className={styles.checkbox}
        label="paste my name"
        data-cy="feedbackName"
      />
      <ProductRating value={rating} onChange={handleChange} />
      <TextField
        sx={{ mb: 4 }}
        value={comment}
        onChange={handleCommentChange}
        label="Comment"
        placeholder={`Min comment length: ${MIN_COMMENT_LENGTH}`}
        variant="outlined"
        fullWidth
        multiline
        minRows={5}
        maxRows={5}
        data-cy="feedbackComment"
      />
      {isLoading ? (
        <LoadingButton label="Loading..." isFullWidth={true} />
      ) : (
        <Button
          fullWidth
          variant="contained"
          onClick={prepareToSend}
          disabled={!isFinished}
          data-cy="feedbackConfirmationBtn"
        >
          Add feedback
        </Button>
      )}
    </Box>
  );
};

FeedbackForms.defaultProps = {
  isLoading: false,
};

FeedbackForms.propTypes = {
  sendForm: PropTypes.func.isRequired,
  isLoading: PropTypes.bool,
};

export default FeedbackForms;
