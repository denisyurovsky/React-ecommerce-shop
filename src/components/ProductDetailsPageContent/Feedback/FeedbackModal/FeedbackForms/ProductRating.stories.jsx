import React from 'react';

import ProductRating from './ProductRating';

export default {
  title: 'rating',
  component: ProductRating,
};

const Template = (args) => <ProductRating {...args} />;

export const Default = Template.bind({});
Default.args = {
  onChange: () => {},
  value: 4,
};

export const Trio = Template.bind({});
Default.args = {
  onChange: () => {},
  value: 3,
};
export const Duo = Template.bind({});
Default.args = {
  onChange: () => {},
  value: 2,
};
