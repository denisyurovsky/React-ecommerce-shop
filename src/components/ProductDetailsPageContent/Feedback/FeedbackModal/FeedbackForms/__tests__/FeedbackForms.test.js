import userEvent from '@testing-library/user-event';
import React from 'react';

import usersDTO from '../../../../../../test-utils/dto/usersDto';
import render, {
  fireEvent,
  screen,
} from '../../../../../../test-utils/renderWith';
import FeedbackForms from '../FeedbackForms';
import getStarByNumber from '../getStarByNumber';

const testRating = 4;
const testText =
  'Sint esse elit id magna qui aliquip commodo aliquip esse amet velit esse ut cupidatat.';

const fillForms = (rating, text) => {
  const comment = screen.getByLabelText('Comment');

  fireEvent.click(getStarByNumber(rating), { bubbles: true });
  userEvent.type(comment, text);
};

describe('FeedbackForms snapshot:', () => {
  it('should render a valid snapshot', () => {
    const { asFragment } = render(<FeedbackForms sendForm={jest.fn()} />);

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Callback functionality', () => {
  const callback = jest.fn();
  const sendData = { rating: testRating, comment: testText, name: 'Anonymous' };
  const sendButton = ['button', { name: /Add feedback/i }];
  const checkboxName = ['checkbox', { name: /pasteName/i }];

  beforeEach(() => {
    render(<FeedbackForms sendForm={callback} />, { user: usersDTO[3] });
    fillForms(testRating, testText);
  });

  it('should be able to send form', () => {
    userEvent.click(screen.getByRole(...sendButton));
    expect(callback).toHaveBeenCalledTimes(1);
  });

  it('should be called with anonymous if no name is provided', () => {
    userEvent.click(screen.getByRole(...sendButton));
    expect(callback).toHaveBeenCalledWith(sendData);
  });

  it('should be called with anonymous by checkbox', () => {
    const checkbox = screen.getByRole(...checkboxName);

    userEvent.click(checkbox);
    expect(checkbox).toBeChecked();

    userEvent.click(checkbox);
    expect(checkbox).not.toBeChecked();

    userEvent.click(screen.getByRole(...sendButton));
    expect(callback).toHaveBeenCalledWith(sendData);
  });

  it('should be called with userName', () => {
    const checkbox = screen.getByRole(...checkboxName);

    userEvent.click(checkbox);
    expect(checkbox).toBeChecked();

    userEvent.click(screen.getByRole(...sendButton));
    expect(callback).toHaveBeenCalledWith({ ...sendData, name: 'Joe Doe' });
  });

  it('should be called with provided name', () => {
    const name = 'ivan';

    userEvent.type(screen.getByLabelText('Enter your name'), name);
    userEvent.click(screen.getByRole(...sendButton));

    expect(callback).toHaveBeenCalledWith({ ...sendData, name: name });
  });
});

describe('Loading state', () => {
  it('should display loading', () => {
    render(<FeedbackForms isLoading={true} sendForm={jest.fn()} />);

    expect(
      screen.getByRole('button', { name: /loading/i })
    ).toBeInTheDocument();
  });
});
