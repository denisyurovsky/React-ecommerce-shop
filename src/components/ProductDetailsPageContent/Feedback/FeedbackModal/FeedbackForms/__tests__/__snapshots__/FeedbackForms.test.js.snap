// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackForms snapshot: should render a valid snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-15md48e-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="mui-1"
        id="mui-1-label"
      >
        Enter your name
      </label>
      <div
        class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-ghsjzk-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-invalid="false"
          autocomplete="off"
          class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
          id="mui-1"
          placeholder="Anonymous"
          type="text"
          value=""
        />
      </div>
    </div>
    <legend
      class="MuiFormLabel-root MuiFormLabel-colorPrimary css-u4tvz2-MuiFormLabel-root"
    >
      Rating
    </legend>
    <div
      class="container MuiBox-root css-0"
    >
      <span
        class="MuiRating-root MuiRating-sizeMedium css-iljtu2-MuiRating-root"
      >
        <label
          class="css-dqr9h-MuiRating-label"
          for="mui-2"
        >
          <span
            class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarRateIcon"
              focusable="false"
              style="opacity: 0.55;"
              viewBox="0 0 24 24"
            >
              <path
                d="M14.43 10 12 2l-2.43 8H2l6.18 4.41L5.83 22 12 17.31 18.18 22l-2.35-7.59L22 10z"
              />
            </svg>
          </span>
          <span
            class="MuiRating-visuallyHidden"
          >
            1 Star
          </span>
        </label>
        <input
          class="MuiRating-visuallyHidden"
          id="mui-2"
          name="hover-feedback"
          type="radio"
          value="1"
        />
        <label
          class="css-dqr9h-MuiRating-label"
          for="mui-3"
        >
          <span
            class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarRateIcon"
              focusable="false"
              style="opacity: 0.55;"
              viewBox="0 0 24 24"
            >
              <path
                d="M14.43 10 12 2l-2.43 8H2l6.18 4.41L5.83 22 12 17.31 18.18 22l-2.35-7.59L22 10z"
              />
            </svg>
          </span>
          <span
            class="MuiRating-visuallyHidden"
          >
            2 Stars
          </span>
        </label>
        <input
          class="MuiRating-visuallyHidden"
          id="mui-3"
          name="hover-feedback"
          type="radio"
          value="2"
        />
        <label
          class="css-dqr9h-MuiRating-label"
          for="mui-4"
        >
          <span
            class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarRateIcon"
              focusable="false"
              style="opacity: 0.55;"
              viewBox="0 0 24 24"
            >
              <path
                d="M14.43 10 12 2l-2.43 8H2l6.18 4.41L5.83 22 12 17.31 18.18 22l-2.35-7.59L22 10z"
              />
            </svg>
          </span>
          <span
            class="MuiRating-visuallyHidden"
          >
            3 Stars
          </span>
        </label>
        <input
          class="MuiRating-visuallyHidden"
          id="mui-4"
          name="hover-feedback"
          type="radio"
          value="3"
        />
        <label
          class="css-dqr9h-MuiRating-label"
          for="mui-5"
        >
          <span
            class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarRateIcon"
              focusable="false"
              style="opacity: 0.55;"
              viewBox="0 0 24 24"
            >
              <path
                d="M14.43 10 12 2l-2.43 8H2l6.18 4.41L5.83 22 12 17.31 18.18 22l-2.35-7.59L22 10z"
              />
            </svg>
          </span>
          <span
            class="MuiRating-visuallyHidden"
          >
            4 Stars
          </span>
        </label>
        <input
          class="MuiRating-visuallyHidden"
          id="mui-5"
          name="hover-feedback"
          type="radio"
          value="4"
        />
        <label
          class="css-dqr9h-MuiRating-label"
          for="mui-6"
        >
          <span
            class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarRateIcon"
              focusable="false"
              style="opacity: 0.55;"
              viewBox="0 0 24 24"
            >
              <path
                d="M14.43 10 12 2l-2.43 8H2l6.18 4.41L5.83 22 12 17.31 18.18 22l-2.35-7.59L22 10z"
              />
            </svg>
          </span>
          <span
            class="MuiRating-visuallyHidden"
          >
            5 Stars
          </span>
        </label>
        <input
          class="MuiRating-visuallyHidden"
          id="mui-6"
          name="hover-feedback"
          type="radio"
          value="5"
        />
        <label
          class="MuiRating-label MuiRating-pristine css-dqr9h-MuiRating-label"
        >
          <input
            class="MuiRating-visuallyHidden"
            id="hover-feedback-empty"
            name="hover-feedback"
            type="radio"
            value=""
          />
          <span
            class="MuiRating-visuallyHidden"
          >
            Empty
          </span>
        </label>
      </span>
      <legend
        class="MuiTypography-root MuiTypography-body1 css-1gvla80-MuiTypography-root"
      />
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-15md48e-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="mui-7"
        id="mui-7-label"
      >
        Comment
      </label>
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <textarea
          aria-invalid="false"
          class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
          id="mui-7"
          placeholder="Min comment length: 30"
          style="height: 0px; overflow: hidden;"
        />
        <textarea
          aria-hidden="true"
          class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
          readonly=""
          style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
          tabindex="-1"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-173wfxe"
          >
            <span>
              Comment
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root Mui-disabled  css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
      disabled=""
      tabindex="-1"
      type="button"
    >
      Add feedback
    </button>
  </div>
</DocumentFragment>
`;
